server/store/user_store.go

```diff
@@ -278,3 +288,86 @@ func (index UserIndex) GetMattermostUserIDs() []string {
 
 	return result
 }
+
+// Get the check user status function
+func GetCheckUserStatus(store Store, logger bot.Logger, mattermostUserID string) func() bool {
+	return func() bool {
+		user, err := store.LoadUser(mattermostUserID)
+		if err != nil {
+			logger.Errorf("Not able to load the user %s. error: %s", mattermostUserID, err.Error())
+			return false
+		}
+
+		// Check if the user is marked as inactive
+		if user.OAuth2Token == nil {
+			return false
+		}
+
+		return true
+	}
+}
+
+// Get the change user status function
+func GetChangeUserStatus(store Store, logger bot.Logger, mattermostUserID string, poster bot.Poster) func(error) {
+	return func(err error) {
+		if err == nil {
+			return
+		}
+
+		if !strings.Contains(err.Error(), ErrorRefreshTokenExpired) {
+			return
+		}
+
+		user, err := store.LoadUser(mattermostUserID)
+		if err != nil {
+			logger.Errorf("Not able to load the user %s. error: %s", mattermostUserID, err.Error())
+			return
+		}
+
+		// Mark the user as inactive
+		user.OAuth2Token = nil
+		if err = store.StoreUser(user); err != nil {
+			logger.Errorf("Not able to store the user %s. error: %s", mattermostUserID, err.Error())
+			return
+		}
+
+		if _, err := poster.DM(mattermostUserID, ErrorUserInactive); err != nil {
+			logger.Errorf("Not able to DM the user %s. error: %s", mattermostUserID, err.Error())
+			return
+		}
+	}
+}
+
+// refreshAndStoreToken checks whether the current access token is expired or not. If it is,
+// then it refreshes the token and stores the new pair of access and refresh tokens in kv store.
+func RefreshAndStoreToken(store Store, token *oauth2.Token, oconf *oauth2.Config, mattermostUserID string) (*oauth2.Token, error) {
+	// If there is only five minutes left for the token to expire, we are refreshing the token.
+	// We don't want the token to expire between the time when we decide that the old token is valid
+	// and the time at which we create the request. We are handling that by not letting the token expire.
+	if time.Until(token.Expiry) > 5*time.Minute {
+		return token, nil
+	}
+
+	src := oconf.TokenSource(context.Background(), token)
+	newToken, err := src.Token() // this actually goes and renews the tokens
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get the new refreshed token")
+	}
+
+	if newToken.AccessToken != token.AccessToken {
```

Comment:
We should also make sure we store the token if the refresh token changes. I would say it's unlikely that it's implemented on Microsoft's side this way, but the OAuth spec does not declare a specific behavior of how this should work, and Microsoft can change their implementation at any time, so we should be defensive of this. Just to make sure we don't need to revisit this, I think we should check both.

```suggestion
	if newToken.AccessToken != token.AccessToken || newToken.RefreshToken != token.RefreshToken {
```